### PR TITLE
Enable Sink and Transformer to be URIs in the spec.

### PR DESCRIPTION
**Pull request by n3wscott**
This is work to support Channel.

### CODE REVIEW

File: config/300-pullsubscription.yaml [hunk 1]
```diff
@@ -80,13 +80,13 @@ spec:
               type: string
               description: "ID of the Cloud Pub/Sub Topic to Subscribe to. It must be in the form of the unique identifier within the project, not the entire name. E.g. it must be 'laconia', not 'projects/my-gcp-project/topics/laconia'."
             ackDeadline:
-              type: string
+              type: integer
```
Comment (vaikas-google) [inline config/300-pullsubscription.yaml]: this seems wrong? see the comments about units in the description?

File: config/300-pullsubscription.yaml [hunk 2]
```diff
@@ -80,13 +80,13 @@ spec:
...
               description:  "The default maximum time after a subscriber receives a message before the subscriber should acknowledge the message. Defaults to `30s`. Valid time units are `s`, `m`, `h`."
             retainAckedMessages:
               type: boolean
               description: "Whether to retain acknowledged messages. If true, acknowledged messages will not be expunged until they fall out of the RetentionDuration window."
             retentionDuration:
-              type: string
+              type: integer
```
Comment (vaikas-google) [inline config/300-pullsubscription.yaml]: ditto

File: pkg/apis/pubsub/v1alpha1/pull_subscription_types.go [hunk 3]
```diff
@@ -94,22 +94,27 @@ type PullSubscriptionSpec struct {
 	// +optional
 	RetentionDuration *time.Duration `json:"retentionDuration,omitempty"`
 
-	// Sink is a reference to an object that will resolve to a domain name to
-	// use as the sink.
+	// Sink is a reference to an object that will resolve to a domain name or a
+	// URI directly to use as the sink.
 	// +optional
-	Sink *corev1.ObjectReference `json:"sink,omitempty"`
+	Sink *Destination `json:"sink,omitempty"`
```
Comment (vaikas-google) [inline pkg/apis/pubsub/v1alpha1/pull_subscription_types.go]: Why is this optional? Just curious.

Comment (vaikas-google) [reply pkg/apis/pubsub/v1alpha1/pull_subscription_types.go]: Looking at the validation code, doesn't seem so optional? :)

File: pkg/apis/pubsub/v1alpha1/pull_subscription_validation.go [hunk 4]
```diff
@@ -74,6 +74,23 @@ func (current *PullSubscriptionSpec) Validate(ctx context.Context) *apis.FieldEr
 	return errs
 }
 
+func validateDestination(dest *Destination) *apis.FieldError {
+	if dest == nil {
+		return apis.ErrMissingField(apis.CurrentField)
```
Comment (vaikas-google) [inline pkg/apis/pubsub/v1alpha1/pull_subscription_validation.go]: this seems to make sink not optional.

File: pkg/reconciler/pullsubscription/pullsubscription.go [hunk 5]
```diff
@@ -191,21 +193,28 @@ func (c *Reconciler) reconcile(ctx context.Context, source *v1alpha1.PullSubscri
 		return nil
 	}
 
-	sinkURI, err := duck.GetSinkURI(ctx, c.DynamicClientSet, source.Spec.Sink, source.Namespace)
-	if err != nil {
-		source.Status.MarkNoSink("NotFound", "")
-		return err
+	// Sink is required. // TODO: this is a problem for channels.
```
Comment (vaikas-google) [inline pkg/reconciler/pullsubscription/pullsubscription.go]: why is this a problem for channels? Is there something that needs to be done for channels? perhaps create an issue for it?

File: pkg/apis/pubsub/v1alpha1/pull_subscription_types.go (hunk 3 above)
Comment (n3wscott) [reply pkg/apis/pubsub/v1alpha1/pull_subscription_types.go]: nice catch! it should not be optional.

File: config/300-pullsubscription.yaml (hunk 1 above)
Comment (n3wscott) [reply config/300-pullsubscription.yaml]: you are correct. It was half wired wrong... fixing.

File: pkg/reconciler/pullsubscription/pullsubscription.go (hunk 5 above)
Comment (n3wscott) [reply pkg/reconciler/pullsubscription/pullsubscription.go]: I was wrong, I have a plan. Todo removed.

File: pkg/apis/pubsub/v1alpha1/pull_subscription_types.go (hunk 3 above)
Comment (vaikas-google) [reply pkg/apis/pubsub/v1alpha1/pull_subscription_types.go]: If it should not be optional, why not:
Sink Destination `json:"sink"`

File: pkg/apis/pubsub/v1alpha1/pull_subscription_types.go [hunk 6]
```diff
@@ -89,27 +88,51 @@ type PullSubscriptionSpec struct {
...
+	// Sink is a reference to an object that will resolve to a domain name or a
+	// URI directly to use as the sink.
+	Sink *Destination `json:"sink,omitempty"`
 
 	// Transformer is a reference to an object that will resolve to a domain
-	// name to use as the transformer.
+	// name or a URI directly to use as the transformer or a URI directly.
 	// +optional
-	Transformer *corev1.ObjectReference `json:"transformer,omitempty"`
+	Transformer *Destination `json:"transformer,omitempty"`
```
Comment (vaikas-google) [inline pkg/apis/pubsub/v1alpha1/pull_subscription_types.go]: this makes sense to be a pointer.

Comment (n3wscott) [reply pkg/apis/pubsub/v1alpha1/pull_subscription_types.go]: Now sink is a pointer.

Comment (vaikas-google) [reply pkg/apis/pubsub/v1alpha1/pull_subscription_types.go]: it was supposed to be paired with the comment about Sink not being a pointer, but this makes sense being a pointer. The comments were supposed to be linked but whatevs. I was trying to make sure you don't change them both to a non-pointer :)

**Pull request merged**